sopel/irc/modes.py

```diff
@@ -0,0 +1,234 @@
+    """Exception when the mode's type management is not configured."""
+    def __init__(self, mode: str, letter: str) -> None:
+        message = 'Type {mode} for mode {letter} is not properly configured.'
+        super().__init__(message.format(mode=mode, letter=letter))
+
+
+def _modes_is_added(modestring: str
+                    ) -> Generator[Tuple[str, bool], None, None]:
+    is_added = True
+    for char in modestring:
+        if char in '+-':
+            is_added = char == '+'
+            continue
+        yield (char, is_added)
+
+
+ModeString = namedtuple('ModeString', [
+    'modes',
+    'privileges',
+    'ignored_modes',
+    'leftover_params',
+])
+
+
+class ModeMessage:
+    """Modestring parser for IRC's ``MODE`` messages for channel modes."""
+    PRIVILEGES: Set[str] = {
+        "v",  # VOICE
+        "h",  # HALFOP
+        "o",  # OP
+        "a",  # ADMIN
+        "q",  # OWNER
+        "y",  # OPER
+        "Y",  # OPER
+    }
+    """Set of user privileges used by default."""
+
+    def __init__(
+        self,
+        chanmodes: Dict[str, Tuple[str, ...]],
+        type_params: Dict[str, str] = DEFAULT_MODETYPE_PARAM_CONFIG,
+        privileges: Set[str] = PRIVILEGES,
+    ) -> None:
+        self.chanmodes: Dict[str, Tuple[str, ...]] = chanmodes
+        """Map of mode types (``str``) with their list of modes (``tuple``).
+
+        This map should come from ``ISUPPORT``, usually through
+        :attr:`bot.isupport.CHANMODES <sopel.irc.isupport.ISupport.CHANMODES>`.
+        """
+        self.type_params = type_params
+        """Map of mode types (``str``) with their param requirements.
+
+        This map default to :data:`DEFAULT_MODETYPE_PARAM_CONFIG`.
+        """
+        self.privileges = privileges
+        """Set of valid user privileges.
+
+        This set should come from ``ISUPPORT``, usually through
+        :attr:`bot.isupport.PREFIX <sopel.irc.isupport.ISupport.PREFIX>`.
+
+        If a server doesn't advertise its prefixes for user privileges,
+        :attr:`PRIVILEGES` can be used as a default value.
+        """
+
+    def get_mode_type(self, mode: str) -> str:
+        """Retrieve the type of ``mode``.
+
+        :raise ModeTypeUnknown: if the mode's type cannont be determined
+        :return: the mode's type as defined by :attr:`chanmodes`
+
+        ::
+
+            >>> mm = ModeMessage({'A': tuple('beI'), 'B': tuple('k')}, {})
+            >>> mm.get_mode_type('b')
+            'A'
+            >>> mm.get_mode_type('k')
+            'B'
+
+        This method will raise a :exc:`ModeTypeUnknown` if the mode is unknown,
+        including the case where ``mode`` is actually a user privilege.
+        (such as ``v``)
+        """
+        for letter, modes in self.chanmodes.items():
+            if mode in modes:
+                return letter
+        raise ModeTypeUnknown(mode)
+
+    def get_mode_info(
+        self,
+        mode: str,
+        is_added: bool
```

Comment:
```suggestion
        is_added: bool,
```

We should have a linter rule that enforces such things, shouldn't we?